drgn_tools/lsmod.py

```diff
@@ -0,0 +1,167 @@
+        table_value.append(["PARAMETER", "ADDRESS", "TYPE", "VALUE"])
+        integer_params = {
+            "byte": "unsigned char *",
+            "short": "short *",
+            "ushort": "unsigned short *",
+            "int": "int *",
+            "uint": "unsigned int *",
+            "long": "long *",
+            "ulong": "unsigned long *",
+        }
+        for i in range(0, mod.obj.num_kp):
+            parm_valf = ""
+            parm_val = ""
+            kp = mod.obj.kp[i]
+            try:
+                param_type = prog.symbol(kp.ops.get).name
+            except LookupError:
+                continue
+
+            if param_type.startswith("param_get_"):
+                param_type = param_type[len("param_get_") :]
+                if param_type in integer_params:
+                    parm_valf_address = cast(
+                        integer_params[param_type], kp.arg
+                    )
+                    try:
+                        parm_valf = parm_valf_address[0].value_()
+                    except FaultError:
+                        parm_valf = "(page fault)"
+                        continue
+                elif "bool" in param_type:
+                    param_type = "bool"
+                    parm_valf_address = cast("bool *", kp.arg)
+                    parm_val = Object(
+                        prog, "bool", address=parm_valf_address.value_()
+                    ).value_()
+                    if parm_val == 1:
+                        parm_valf = "Y"
+                    else:
+                        parm_valf = "N"
+                elif "charp" in param_type:
+                    param_type = "charp"
+                    value = Object(prog, "char **", value=kp.arg)
+                    if value[0]:
+                        parm_valf = escape_ascii_string(value[0].string_())
+                    else:
+                        parm_valf = "(null)"
+                elif "string" in param_type:
+                    param_type = "string"
+                    try:
+                        if escape_ascii_string(kp.str.string.string_()):
+                            parm_valf = escape_ascii_string(
+                                kp.str.string.string_()
+                            )
+                        else:
+                            parm_valf = "(null)"
+                    except FaultError:
+                        parm_valf = "(page fault)"
+                        continue
+            elif "array" in param_type:
+                param_type = "array"
+                parm_valf = str(
+                    int(str(prog.read(kp.arg.value_(), 1).hex()), 16)
+                )
+            table_value.append(
+                [
+                    str(kp.name.string_().decode("utf-8")),
+                    str(ker_parm),
+                    param_type,
+                    str(parm_valf),
+                ]
+            )
+            ker_parm = hex(int(ker_parm, 16) + int(40))
```

Comment:
I'm assuming that 40 is the size of `struct kernel_param` and you're adjusting the hex value to point to the next parameter?

This is contrary to the entire point of drgn: it knows the size of structures, and it can tell you the address of objects. Rather than maintaining your own variable, and hard-coding structure sizes, let's have drgn maintain everything for us.

Please delete the `kpr_parm` variable entirely. When you need the hex address of the `struct kernel_param`, you can do `hex(kp.address_of_())`.

Reply:
added string "array_length" to the output showing the length of the array

MODULE NAME:    parport_pc                                    
PARAM COUNT: 5                                                
ADDRESS    : 0xffffffffc0327390                               
PARAMETER  ADDRESS             TYPE   VALUE                   
init_mode  0xffffffffc03268a0  charp  (null)                  
dma        0xffffffffc03268c0  array  array_length: 16        
irq        0xffffffffc03268e0  array  array_length: 16        
io_hi      0xffffffffc0326900  array  array_length: 17        
io         0xffffffffc0326920  array  array_length: 17